torchx/schedulers/azure_batch_scheduler.py

```diff
@@ -0,0 +1,338 @@
+
+from torchx.specs.api import (
+    AppDef,
+    AppState,
+    macros,
+    runopts,
+    RetryPolicy,
+    Role,
+)
+
+from azure.batch import BatchServiceClient
+
+import azure.batch.models as batchmodels
+from azure.batch import BatchServiceClient
+from azure.batch.batch_auth import SharedKeyCredentials
+import azure.batch.models as batchmodels
+
+
+
+@dataclass
+class AzureBatchJob:
+    job_id: str
+    pool_id: str
+    job: batchmodels.JobAddParameter
+    tasks: List[batchmodels.TaskAddParameter]
+
+    def __str__(self) -> str:
+        return yaml.dump({"job": self.job, "tasks": self.tasks})
+
+    def __repr__(self) -> str:
+        return str(self)
+
+
+class AzureBatchOpts(TypedDict, total=False):
+    batch_container_registry: Optional[str]
+    batch_container_username: Optional[str]
+    batch_container_password: Optional[str]
+    batch_pool_id: Optional[str]
+
+    
+JOB_STATE: Dict[str, AppState] = {
+    # TODO: instead of state to state mapping, use Job transition states (prev+current) to capture status
+    batchmodels.JobState.active: AppState.RUNNING,
+    batchmodels.JobState.completed: AppState.SUCCEEDED,
+    batchmodels.JobState.deleting: AppState.UNKNOWN,
+    batchmodels.JobState.disabled: AppState.UNKNOWN,
+    batchmodels.JobState.disabling: AppState.UNKNOWN,
+    batchmodels.JobState.enabling: AppState.PENDING,
+    batchmodels.JobState.terminating: AppState.UNKNOWN,
+}
+
+def _as_metadata_model(name:str, value:str) -> batchmodels.MetadataItem:
+    # pyre-ignore[16]
+    return batchmodels.MetadataItem(name=name, value=value)
+
+class AzureBatchScheduler(Scheduler[AzureBatchOpts]):
+    """
+    AzureBatchScheduler is a TorchX scheduling interface to Azure Batch.
+
+    .. code-block:: bash
+
+        $ pip install torchx
+        $ torchx run --scheduler azure_batch --scheduler_args image_repo=ghcr.io/pytorch/torchx --image alpine:latest --msg hello
+        azure_batch://torchx_user/1234
+        $ torchx describe azure_batch://torchx_user/1234
+        ...
+
+    **Config Options**
+
+    .. runopts::
+        class: torchx.schedulers.azure_batch_scheduler.create_scheduler
+
+
+    **Compatibility**
+
+    .. compatibility::
+        type: scheduler
+        features:
+            cancel: true
+            logs: false
+            distributed: true
+            describe: true
+            workspaces: false
+            mounts: false
+            elasticity: false
+    """
+
+    def __init__(
+        self,
+        session_name: str,
+        batch_account_name: Optional[str],
+        batch_account_key: Optional[str],
+        batch_account_url: Optional[str],
+    ) -> None:
+        super().__init__("azure_batch", session_name)
+        self.batch_account_name = batch_account_name
+        self.batch_account_key = batch_account_key
+        self.batch_account_url = batch_account_url
+
+    def _build_client(self) -> BatchServiceClient:
+        credentials = SharedKeyCredentials(
```

Comment:
fwiw, for AWS, the boto3 client grabs the default credentials (as configured in `~/.aws/credentials` or from instance profile (if on an EC2 instance). Does a similar concept not exist in Azure? If it does we should follow the same strategy to deal with credentials.

For example, if a user using `aws_batch` scheduler wanted to change to a different credential (either to access a different account or to switch to a different IAM role), they would change the creds provider in the `~/.aws` configuration rather than having to pass this information to torchx. This is much clearer to the users since they can use the vanilla AWS credentials chain  rather than having to pass parameters to torchx.